applications/machine_learning/src/modules/usb_state.c

```diff
@@ -8,15 +8,108 @@
 
 #define MODULE usb_state
 #include <caf/events/module_state_event.h>
+#include <caf/events/power_manager_event.h>
 
 #include <logging/log.h>
 LOG_MODULE_REGISTER(MODULE, CONFIG_ML_APP_USB_STATE_LOG_LEVEL);
 
+/** @brief Peer states. */
+enum usb_state {
+	USB_STATE_DISCONNECTED, /**< Cable is not attached. */
+	USB_STATE_POWERED,      /**< Cable attached but no communication. */
+	USB_STATE_ACTIVE,       /**< Cable attached and data is exchanged. */
+	USB_STATE_SUSPENDED,    /**< Cable attached but port is suspended. */
+	USB_STATE_ERROR         /**< USB error */
+};
+
+
+/* Current state of the USB */
+static enum usb_state current_state = USB_STATE_DISCONNECTED;
+
+
+static void device_status(enum usb_dc_status_code status, const uint8_t *param)
+{
+	static enum usb_state before_suspend;
+	enum usb_state new_state = current_state;
+
+	switch (status) {
+	case USB_DC_ERROR:
+		new_state = USB_STATE_ERROR;
+		module_set_state(MODULE_STATE_ERROR);
+		break;
+	case USB_DC_RESET:
+		__ASSERT_NO_MSG(current_state != USB_STATE_DISCONNECTED);
+		new_state = USB_STATE_ACTIVE;
+		break;
+	case USB_DC_CONNECTED:
+		new_state = USB_STATE_POWERED;
+		break;
+	case USB_DC_CONFIGURED:
+		__ASSERT_NO_MSG(current_state != USB_STATE_DISCONNECTED);
+		new_state = USB_STATE_ACTIVE;
+		break;
+	case USB_DC_DISCONNECTED:
+		new_state = USB_STATE_DISCONNECTED;
+		break;
+	case USB_DC_SUSPEND:
+		if (current_state == USB_STATE_DISCONNECTED) {
+			/* Due to the way USB driver and stack are written
```

Comment:
I think that one should be fixed already. Maybe we could consider removing workaround ?

Reply:
(also in nRF Desktop)

Reply:
Can we create separated issue to fix that? I do not like to fix this and desktop here. This functionality would not break anything.

Reply:
This one must be triple checked btw.

Maybe we should at the same time think how to move usb_state to caf. I know this can be tricky as nrf_desktop uses hid class instances but maybe we can export the common part and find a nice way of integrating hid class support handling our hid_report events with it. Will create a story for it.

Reply:
Added story for the fix.